pkg/velerorunner/registry.go

```diff
@@ -0,0 +1,526 @@
+		registryResources.CredSecret = newSecret
+		err = client.Create(context.TODO(), newSecret)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	registryResources.CredSecret = foundSecret
+	if !t.equalsRegistryCredentialsSecret(newSecret, foundSecret) {
+		t.updateRegistryCredentialsSecret(client, foundSecret)
+		err = client.Update(context.TODO(), foundSecret)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Build a credentials Secret as desired for the source cluster.
+func (t *Task) buildRegistryCredentialsSecret(client k8sclient.Client) (*kapi.Secret, error) {
+	secret := &kapi.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:       t.PlanResources.MigPlan.GetCorrelationLabels(),
+			GenerateName: t.PlanResources.MigPlan.GetName() + "-registry-",
+			Namespace:    VeleroNamespace,
+		},
+	}
+	err := t.updateRegistryCredentialsSecret(client, secret)
+	return secret, err
+}
+
+// Update a Registry credentials secret as desired for the specified cluster.
+func (t *Task) updateRegistryCredentialsSecret(client k8sclient.Client, secret *kapi.Secret) error {
+	credSecret, err := t.PlanResources.MigStorage.Spec.BackupStorageConfig.GetCredsSecret(t.Client)
+	if err != nil {
+		return err
+	}
+	if credSecret == nil {
+		return migapi.CredSecretNotFound
+	}
+	secret.Data = map[string][]byte{
+		"access_key": []byte(credSecret.Data[migapi.AwsAccessKeyId]),
+		"secret_key": []byte(credSecret.Data[migapi.AwsSecretAccessKey]),
+	}
+	return nil
+}
+
+// Get an existing credentials Secret on the source cluster.
+func (t *Task) getRegistryCredentialsSecret(client k8sclient.Client) (*kapi.Secret, error) {
+	list := kapi.SecretList{}
+	labels := t.PlanResources.MigPlan.GetCorrelationLabels()
+	err := client.List(
+		context.TODO(),
+		k8sclient.MatchingLabels(labels),
+		&list)
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Items) > 0 {
+		return &list.Items[0], nil
+	}
+
+	return nil, nil
+}
+
+// Determine if two registry credentials secrets are equal.
+// Returns `true` when equal.
+func (t *Task) equalsRegistryCredentialsSecret(a, b *kapi.Secret) bool {
+	return reflect.DeepEqual(a.Data, b.Data)
+}
+
+// Ensure the imagestream for the migration registry on the specified cluster has been created
+func (t *Task) ensureRegistryImageStream(client k8sclient.Client, registryResources *MigrationRegistryResources) error {
+	registrySecret := registryResources.CredSecret
+	if registrySecret == nil {
+		return errors.New("migration registry credentials not found")
+	}
+	newImageStream, err := t.buildRegistryImageStream(registrySecret.GetName())
+	if err != nil {
+		return err
+	}
+	foundImageStream, err := t.getRegistryImageStream(client)
+	if err != nil {
+		return err
+	}
+	if foundImageStream == nil {
+		registryResources.ImageStream = newImageStream
+		err = client.Create(context.TODO(), newImageStream)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	registryResources.ImageStream = foundImageStream
```

Comment:
Wondering if this is intended. it looks like `registryResources.ImageStream` gets set to newImageStream above, but that will always be rewritten by this value unless the create errors out.

Reply:
Note the `return nil` at the end of that block. If it doesn't yet exist, it will return after setting it to newImageStream or it will error out. Otherwise, it found one already and sets it to that found value here. The backup creation follows the same logic.

Reply:
Ah, missed that. Thanks for pointing that out.